src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java

```diff
@@ -77,6 +79,70 @@ public PostgreSqlDatabaseDialect(AbstractConfig config) {
     super(config, new IdentifierRules(".", "\"", "\""));
   }
 
+  @Override
+  public Connection getConnection() throws SQLException {
+    Connection result = super.getConnection();
+    synchronized (this) {
+      if (maxIdentifierLength <= 0) {
+        maxIdentifierLength = computeMaxIdentifierLength(result);
+      }
+    }
+    return result;
+  }
+
+  static int computeMaxIdentifierLength(Connection connection) {
+    String warningMessage = "Unable to query database for maximum table name length; "
+        + "the connector may fail to write to tables with long names";
+    // https://stackoverflow.com/questions/27865770/how-long-can-postgresql-table-names-be/27865772#27865772
+    String nameLengthQuery = "SELECT length(repeat('1234567890', 1000)::NAME);";
+    
+    int result;
+    try (ResultSet rs = connection.createStatement().executeQuery(nameLengthQuery)) {
+      if (rs.next()) {
+        result = rs.getInt(1);
+        if (result <= 0) {
+          log.warn(
+              "Cannot accommodate maximum table name length of {} as it is not positive; "
+                  + "table name truncation will be disabled, "
+                  + "and the connector may fail to write to tables with long names",
+              result);
+          result = Integer.MAX_VALUE;
+        } else {
+          log.info(
+              "Maximum table name length for database is {} bytes",
+              result
+          );
+        }
+      } else {
+        log.warn(warningMessage);
+        result = Integer.MAX_VALUE;
+      }
+    } catch (SQLException e) {
+      log.warn(warningMessage, e);
+      result = Integer.MAX_VALUE;
+    }
+    return result;
+  }
+
+  @Override
+  public TableId parseTableIdentifier(String fqn) {
+    TableId result = super.parseTableIdentifier(fqn);
+    if (maxIdentifierLength > 0 && result.tableName().length() > maxIdentifierLength) {
+      String newTableName = result.tableName().substring(0, maxIdentifierLength);
```

Comment:
Could we unintentionally cause table name collisions by doing this?

Reply:
Yes, this is definitely possible. @gharris1727 even noted it a couple years ago on the [GitHub issue](https://github.com/confluentinc/kafka-connect-jdbc/issues/675#issuecomment-566337721) linked in the description:

> ... it could cause table collisions, if two topics differ by only their truncated suffixes.

I think the approach here is still reasonable since we are only adopting to behavior that PostgreSQL applies downstream. If users encounter issues (such as collision) because of table name truncation, it's difficult to argue that it's the connector's fault when they'd see the exact same effects if they ran the same DDL/DML statements against the database by hand.

One approach we might consider is adding a configuration property that gives users some control over this situation, including but not limited to these options:
- Fail if the table name is too long
- Truncate if the table is too long
- Attempt to deduce the maximum table length and truncate accordingly, but disable truncation if the table length cannot be determined

I opted for an approach that could be backported as it did not introduce a new configuration property and, IMO, has reasonable behavior. But this is a bit borderline and if you have different thoughts on how we can/should handle this I think it's worth it to have that discussion so that we can (hopefully) get this right the first time.

Reply:
I agree with the approach you have in this PR. If you have a follow-up PR on the docs repo, that should be good enough. This way it's easy to explain to unsuspecting customers.